Modernise Manage Jenkins screen

janfaracik opened the pull request:
### Proposed changelog entries

* Modernise the "Manage Jenkins" screen

### Screenshot

Old
<img width="1110" alt="image" src="https://user-images.githubusercontent.com/43062514/130867385-5ad4f79c-795f-466e-8208-5d34450d146f.png">

New
<img width="1109" alt="image" src="https://user-images.githubusercontent.com/43062514/130867352-29fff9cd-a9ec-411c-bfdd-83b16c9def97.png">

Hover/focus states have been updated to match buttons, text and padding sizes have been tweaked to make page more readable and coherent. 

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

janfaracik committed:
Initial

janfaracik committed:
Update manage-jenkins.less

janfaracik committed:
Update manage-jenkins.less

janfaracik committed:
Merge branch 'master' into revamp-settings

janfaracik committed:
Cleanup

janfaracik committed:
Delete breadcrumb-config-outline.properties

oleg-nenashev reviewed (approved):
Looks fine with Dark Theme too, thanks!

![image](https://user-images.githubusercontent.com/3000480/130947944-ae1c91d1-d206-463f-9806-3274efab3ce1.png)

daniel-beck commented:
Design wise, this looks nice.

---

Could you clarify why you decided to remove the scaling based on window width? This now is always 3 columns wide, rather than scaling from 1 to 4, depending on available space on the screen. This was a deliberate decision so that individual entries are not overly squished, no matter the screen size.

Extremes, before (alt text for description):

> <img width="357" alt="Most narrow, 2 columns" src="https://user-images.githubusercontent.com/1831569/131212315-a6127b39-0c32-4bfa-8824-1bc3afb2a5b5.png">
> <img width="599" alt="Widest, 4 columns" src="https://user-images.githubusercontent.com/1831569/131212317-18a8879e-bfe6-4b6b-90d7-1fbaec68c0d1.png">

Extremes, after:

> <img width="272" alt="With sidepanel, 3 columns, narrow" src="https://user-images.githubusercontent.com/1831569/131212325-58e26372-2ca1-466a-949f-edbcf0878bc1.png">
> <img width="767" alt="With sidepanel, 3 columns, wide" src="https://user-images.githubusercontent.com/1831569/131212329-79929959-c04f-4706-aad4-8a448cb72c2d.png">

---

When the layout moves content below the "side"panel, it previously went to 1 column width, i.e. a basic list. With this PR it's still 3 columns wide. Of course it's a valid question to ask how often this is relevant, but I think recent work even on forms was to make it more mobile/small screen friendlier, so this looks like a step back for that.

> <img width="589" alt="Screenshot 2021-08-28 at 10 56 40" src="https://user-images.githubusercontent.com/1831569/131212458-6f07fd3c-71ff-4192-ac22-b4325a036991.png">

> <img width="448" alt="Screenshot 2021-08-28 at 11 03 24" src="https://user-images.githubusercontent.com/1831569/131212623-e05a1d94-c6c5-4c63-9a83-d901b7c34d67.png">

janfaracik commented:
> Wouldn't it make sense to increase the font size of the sections? Currently the section titles (h2 tag) are much smaller than the heading and texts of the individual buttons.

It was solely a design decision, somewhat inspired by Material Design or iOS - 

![image](https://user-images.githubusercontent.com/43062514/131218824-e3dd849e-a5e8-4987-97e9-90fdcdd7d040.png)

![image](https://user-images.githubusercontent.com/43062514/131218843-f8a38715-cb8f-41c0-adf9-2b181e1bba38.jpeg)

janfaracik commented:
> Design wise, this looks nice.
> 
> Could you clarify why you decided to remove the scaling based on window width? This now is always 3 columns wide, rather than scaling from 1 to 4, depending on available space on the screen. This was a deliberate decision so that individual entries are not overly squished, no matter the screen size.
> 
> Extremes, before (alt text for description):
> 
> > <img alt="Most narrow, 2 columns" width="357" src="https://user-images.githubusercontent.com/1831569/131212315-a6127b39-0c32-4bfa-8824-1bc3afb2a5b5.png">
> > <img alt="Widest, 4 columns" width="599" src="https://user-images.githubusercontent.com/1831569/131212317-18a8879e-bfe6-4b6b-90d7-1fbaec68c0d1.png">
> 
> Extremes, after:
> 
> > <img alt="With sidepanel, 3 columns, narrow" width="272" src="https://user-images.githubusercontent.com/1831569/131212325-58e26372-2ca1-466a-949f-edbcf0878bc1.png">
> > <img alt="With sidepanel, 3 columns, wide" width="767" src="https://user-images.githubusercontent.com/1831569/131212329-79929959-c04f-4706-aad4-8a448cb72c2d.png">
> 
> When the layout moves content below the "side"panel, it previously went to 1 column width, i.e. a basic list. With this PR it's still 3 columns wide. Of course it's a valid question to ask how often this is relevant, but I think recent work even on forms was to make it more mobile/small screen friendlier, so this looks like a step back for that.
> 
> > <img alt="Screenshot 2021-08-28 at 10 56 40" width="589" src="https://user-images.githubusercontent.com/1831569/131212458-6f07fd3c-71ff-4192-ac22-b4325a036991.png">
> 
> > <img alt="Screenshot 2021-08-28 at 11 03 24" width="448" src="https://user-images.githubusercontent.com/1831569/131212623-e05a1d94-c6c5-4c63-9a83-d901b7c34d67.png">

Hm that's behaving a bit differently on my end, I'll restore the original way so that it behaves the correct way 👍

janfaracik committed:
Update manage-jenkins.less

janfaracik committed:
Update manage-jenkins.less

janfaracik committed:
Use app bar styling

janfaracik committed:
Merge branch 'master' into revamp-settings

daniel-beck commented:
FWIW we've generally tried to keep PR diffs minimal to make changes easier to review, cherry-pick, and merge -- even if the overall file format was inconsistent and wrong.

oleg-nenashev commented:
> FWIW we've generally tried to keep PR diffs minimal to make changes easier to review, cherry-pick, and merge -- even if the overall file format was inconsistent and wrong.

This is not a bugfix, and there are no pending pull requests against this screen. I am fine with keeping refactoring integrated though I agree with @daniel-beck in general

timja reviewed (approved):
Lgtm

timja commented:
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

daniel-beck commented:
I'd like an opportunity to test this PR again before it's merged.

res0nance commented:
Tested on I.E

I see this issue
![image](https://user-images.githubusercontent.com/31362124/131663898-ee010e90-f3a7-4344-b93c-b8fb21dda732.png)

Hovering over system log seems to cause the button to overlap?

timja commented:
> I'd like an opportunity to test this PR again before it's merged.

Sure, I tested it on my 27” monitor and it was fine

daniel-beck commented:
Is it deliberate that administrative monitor warnings appear between horizontal bar and "System Configuration" title? Looking at the screenshots I got the impression of horizontal bars and titles as a unit, but they're split up at the first one.

> <img width="1168" alt="Screenshot 2021-09-01 at 14 18 31" src="https://user-images.githubusercontent.com/1831569/131669906-541e6819-66f7-4e43-bbe8-8226bc4966a1.png">

I'm not rejecting it, it's just something I noticed.

janfaracik committed:
Add IE11 margin support

janfaracik commented:
> Tested on I.E
> 
> I see this issue
> ![image](https://user-images.githubusercontent.com/31362124/131663898-ee010e90-f3a7-4344-b93c-b8fb21dda732.png)
> 
> Hovering over system log seems to cause the button to overlap?

It's due to the lack of gap between the buttons sadly - it turns out that IE11 doesn't support `grid-gap` annoyingly so I've added a workaround that adds margins to the buttons for IE11 - should solve that issue for you :)

janfaracik commented:
> Is it deliberate that administrative monitor warnings appear between horizontal bar and "System Configuration" title? Looking at the screenshots I got the impression of horizontal bars and titles as a unit, but they're split up at the first one.
> 
> > <img alt="Screenshot 2021-09-01 at 14 18 31" width="1168" src="https://user-images.githubusercontent.com/1831569/131669906-541e6819-66f7-4e43-bbe8-8226bc4966a1.png">
> 
> I'm not rejecting it, it's just something I noticed.

<img width="1083" alt="image" src="https://user-images.githubusercontent.com/43062514/131704679-1e79f0c1-6f9b-4825-b68f-0e09d44f79a4.png">

Ah it was originally like that, must have broken it by mistake when changing the header. I've restored the OG look :)

janfaracik committed:
Fix alert placement

janfaracik committed:
Merge branch 'master' into revamp-settings

janfaracik committed:
Fix no alert margin

janfaracik committed:
Update app-bar.less

daniel-beck commented:
@janfaracik Thanks for the super quick reaction!

daniel-beck commented:
> @daniel-beck are you wanting to look again?

Yes.

oleg-nenashev commented:
Gentle ping @daniel-beck. There is no time pressure for the next LTS baseline, but it would be awesome to merge it. There are many Web UI PRs coming it, and I'd rather avoid merge conflicts

timja commented:
reminder @daniel-beck.

daniel-beck commented:
I tested the current PR and it works well. My feedback has been addressed fully, thanks for those changes. I like in particular the rounded corners and "active" border around entries on this page. The headers look good IMO in isolation. As far as I can tell (not a CSS expert), the implementation is done well too.

If this is the first part of a larger scale planned redesign involving forms etc., it's a very promising start and I look forward to future changes.

If this is all that's planned, I am not sure the modernization is worth the additional inconsistency with other parts of the Jenkins UI. While a few places with vertical dividers exist (e.g. with nested sidepanel items, _below_ section headers on config forms), they're not used like these are, and the small all-caps headers are a completely new element in Jenkins. I realize this screen has always been different from others, and I contributed to making it even more so in the past, but that was mostly elements that didn't exist elsewhere. The parts that have equivalents elsewhere (headers) remained similar.

I don't feel qualified enough in the UI space to either approve or reject this PR as a whole.

uhafner reviewed (commented):
It looks better than before, so I am fine with merging this PR. 

But I am still unsure about the new design of the section headers. I think the texts are too small and all caps texts are not used in Jenkins up to now. And they are not consistent with the rest of Jenkins UI (they weren't before either). 

E.g. in the configure screens we have:

![Bildschirmfoto 2021-09-13 um 23 20 13](https://user-images.githubusercontent.com/503338/133158767-e9bbb483-1d60-4384-b683-3007ef53d619.png)
  
This looks better IMHO.

fqueiruga commented:
I think this looks great, cheers! Heading consistency is a problem, but I can see this page as a one-off screen regarding some design elements.

fqueiruga commented on war/src/main/less/base/layout-commons.less:
```diff
@@ -76,7 +76,7 @@ body {
 
 #main-panel {
-  padding: 15px 15px 40px 15px;
+  padding: 30px;
```
might as well use `2rem` here

fqueiruga replied:
Also, I'm a bit -1 on this change. This will affect everything, including page footers, etc.

janfaracik replied:
I've changed it to 2rem. The padding will affect every page, but I think it's worth it as it improves the spacing of Jenkins' UI and gives it more balance.

fqueiruga commented on war/src/main/less/modules/app-bar.less:
```diff
@@ -0,0 +1,32 @@
+.jenkins-app-bar {
```
What does app-bar mean? If it's for the configure screen, it should have a component-specific name. We already have CSS for the manage jenkins page.

janfaracik replied:
The app bar represents the heading for the page, it contains the title (in this case Manage Jenkins) with the option of placing components on the far right of it. I wanted to get it merged in with this MR as I plan on making extensive use of it in future design related MRs.

fqueiruga commented on war/src/main/less/modules/app-bar.less:
```diff
@@ -0,0 +1,32 @@
+    margin-bottom: 0;
+  }
+
+  h1, h2, h3 {
+    margin: 0;
+    font-size: 30px;
```
We should style heading sizes using the `.h1`, `.h2` classes, etc. Also, font-size should be in rems in any case.

fqueiruga commented on war/src/main/less/modules/manage-jenkins.less:
```diff
@@ -1,152 +1,175 @@
-.manage-option {
-    border: 1px solid transparent;
-    margin-bottom: 4px;
+.jenkins-settings {
```
What does `.jenkins-settings` mean? Sorry for being cheeky, but if it's just for the manage page I'd just name it like it is, not a generic jenkins settings component.

fqueiruga replied:
I would guess it would apply to the general settings page, or plugin manager for example.

janfaracik replied:
I named it `jenkins-settings` over `jenkins-manage/jenkins-manage-jenkins/manage-jenkins` or something along those lines as it's a touch shorter and more readable. I do recognise that it might not be instantly clear however, I think moving the file into a `pages` package over the current `modules` package could help devs understand that its styles for a specific page rather than a component?

fqueiruga commented on war/src/main/less/modules/manage-jenkins.less:
```diff
@@ -1,152 +1,175 @@
-    border: 1px solid transparent;
-    margin-bottom: 4px;
+.jenkins-settings {
+  &__section-title {
```
I'm always a proponent of using unexploded classnames. I think SCSS's nesting hurts maiteinability. Would you mind using fully-qualified selectors like `jenkins-settings__section-title` or `.jenkins-settings__section?

janfaracik replied:
Done :)

fqueiruga replied:
Thanks!

fqueiruga commented on war/src/main/less/modules/manage-jenkins.less:
```diff
@@ -1,152 +1,175 @@
-    margin: 0 0 0 62px;
-    min-height: 48px;
-}
+    .manage-page__row {
```
This is sort of what I meant, tracking the `.jenkins-settings__section .mange-page__row` to this file would be very quite hard doing file search.

fqueiruga commented on war/src/main/less/modules/manage-jenkins.less:
```diff
@@ -1,152 +1,175 @@
-    &:hover,
-    &:focus-within {
-      border: 1px solid var(--manage-option-border-color);
+    &:first-of-type {
```
I'm fine with leaving pseudoselectors nested.

fqueiruga commented on war/src/main/less/modules/manage-jenkins.less:
```diff
@@ -1,152 +1,175 @@
+  &:hover,
+  &:focus {
+    &::before {
+      background-color: var(--btn-link-bg--hover);
```
Can't remember now, but doesn't this also have variables for button link color and border?

fqueiruga commented on war/src/main/less/modules/manage-jenkins.less:
```diff
@@ -1,152 +1,175 @@
-    float: left;
+  }
+  
+  dl {
```
Is this rule nested under `.jenkins-settings`?

janfaracik replied:
As mentioned above, that styling as now been moved out into its own file under `.jenkins-section`

fqueiruga reviewed (changes requested):
I have a few comments regarding CSS style and naming for the classes. `.jenkins-settings` feels a bit too generic to me.

The only change I can consider a must is revisit the padding of the `#main-panel`. It can have side effects across quite a few plugins and screens.

janfaracik committed:
Update manage-jenkins.less

janfaracik commented:
> If this is the first part of a larger scale planned redesign involving forms etc., it's a very promising start and I look forward to future changes.
> 
> If this is all that's planned, I am not sure the modernization is worth the additional inconsistency with other parts of the Jenkins UI. While a few places with vertical dividers exist (e.g. with nested sidepanel items, _below_ section headers on config forms), they're not used like these are, and the small all-caps headers are a completely new element in Jenkins. I realize this screen has always been different from others, and I contributed to making it even more so in the past, but that was mostly elements that didn't exist elsewhere. The parts that have equivalents elsewhere (headers) remained similar.

I've got some branches on my fork of Jenkins at the moment looking at improving the interface across the service - updating Jenkins' forms definitely plays a large part in that. Current WIP for that below:

![Jenkins Forms](https://user-images.githubusercontent.com/43062514/133934464-7d5cc2aa-365b-4d77-8dbf-f0ef9835a446.png)

> It looks better than before, so I am fine with merging this PR.
> 
> But I am still unsure about the new design of the section headers. I think the texts are too small and all caps texts are not used in Jenkins up to now. And they are not consistent with the rest of Jenkins UI (they weren't before either).

As for the headers, I'm removed the all caps text and increased the size so that it's consistent with the rest of Jenkins.

It looks like so -

<img width="701" alt="image" src="https://user-images.githubusercontent.com/43062514/133934676-465e1ad6-a082-47cb-af4f-450afbab91aa.png">

Hope that addresses your comments :)

janfaracik committed:
Use REM

janfaracik committed:
Move section styling to its own file

fqueiruga commented on war/src/main/less/modules/app-bar.less:
```diff
@@ -0,0 +1,32 @@
+  justify-content: space-between;
+  margin-bottom: 30px;
+  padding-bottom: 30px;
+  border-bottom: 2px solid var(--panel-border-color);
```
nice to see CSS vars <3

fqueiruga commented on war/src/main/less/modules/app-bar.less:
```diff
@@ -0,0 +1,32 @@
+  display: flex;
+  align-items: center;
+  justify-content: space-between;
+  margin-bottom: 30px;
+  padding-bottom: 30px;
```
I would suggest this could be changed to `2rem`. All `30px` instances really.

janfaracik replied:
Done :)

timja commented:
> hudson.model.ManagementLinkTest.links failed

https://github.com/jenkinsci/jenkins/pull/5693/checks?check_run_id=3646260784
I assume related ^^

fqueiruga reviewed (approved):
I'm happy to hear that there are intentions of working on forms. I think it's needed. I think that would be great, especially if all stuff like input border, padding, height, etc is configurable.

Regarding the reusability of the headings... As @daniel-beck commented on https://github.com/jenkinsci/jenkins/pull/5693#issuecomment-918517681, this screen has always be a one-off. I see nothing wrong in having a one-off header just for this page. It can also be tricky to do.

Thanks for the PR! Approving the code based on how it looks, haven't tried it. Some specific feedback to note is that dimensions above 4-6px should probably be in `rem`s IMO. I love the use of existing CSS vars.

Cheers!

timja commented:
Not sure if I'm a fan of the supporting text weight change:

(Looks like the `font-weight` changed from 400 to 600)

Before:

![image](https://user-images.githubusercontent.com/21194782/133976961-2b5c4200-625e-4231-b2c3-48997e8e4c41.png)

After:

![image](https://user-images.githubusercontent.com/21194782/133976991-1b921003-ad60-4817-ac0c-d0303c7d7a4a.png)

Imo it especially doesn't look so good when there's a lot of text:

![image](https://user-images.githubusercontent.com/21194782/133977037-d8b2d109-71b0-42e7-b821-8c17a0ac5b25.png)

any thoughts on that ^?

Apart from that all looks good to me

fqueiruga commented:
I agree @timja, it's a subtitle after all.

timja commented on war/src/main/less/modules/section.less:
```diff
@@ -0,0 +1,108 @@
+  dd {
+    color: var(--text-color);
+    opacity: 0.75;
+    font-weight: 600;
```
this should fix the issue I'm seeing with the text being too bold
```suggestion
    font-weight: 400;
```

janfaracik replied:
Done :)

janfaracik committed:
Use REM, potentially fix test?

janfaracik commented:
> > hudson.model.ManagementLinkTest.links failed
> 
> https://github.com/jenkinsci/jenkins/pull/5693/checks?check_run_id=3646260784
> I assume related ^^

Strange, I ran the test in IntelliJ and all was fine :< I've updated the selector anyhow to use the new class name,  hopefully that'll fix it

timja commented:
> > > hudson.model.ManagementLinkTest.links failed
> > 
> > 
> > [#5693 (checks)](https://github.com/jenkinsci/jenkins/pull/5693/checks?check_run_id=3646260784)
> > I assume related ^^
> 
> Strange, I ran the test in IntelliJ and all was fine :< I've updated the selector anyhow to use the new class name, hopefully that'll fix it

probably cached jelly files in some way, sometimes IntelliJ goes quite crazy with core tests.
Clean build or invalidating caches might fix it.

Sometimes running tests in intellij works fine and sometimes you get non-sensical errors 😢 .

timja reviewed (approved):
looks better now

![image](https://user-images.githubusercontent.com/21194782/133990820-a5769a88-ca77-4d2b-9fa9-c0c20b469f5f.png)

timja commented:
@daniel-beck did you want to take a last look at this?

daniel-beck reviewed (approved):
Looks good. Thanks!

timja commented:
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

dheerajodha commented:
Proposed changelog entries

- Modernise Manage Jenkins screen

can be changed to:

- Modernise the <code>Manage Jenkins</code> screen.

to put emphasis on the screen which was changed.

Bananeweizen commented:
@janfaracik Please don't blindly copy design from your phone. The slider element in your WIP screenshot "Publish results with TM4J") is a rather bad choice for checkbox input elements in an application like Jenkins that is mainly used on operating systems _without_ touch. The metaphor of sliding left and right doesn't fit when most people use a mouse click for it.
